### ISSUE TITLE
getPayloadData.py script broken in CMSSW_12_0_0_pre3

### ISSUE THREAD

**Issue by mmusich** [cms-sw/cmssw#34193]
As can be seen in [CondDB Browser logs](https://cms-conddb.cern.ch/admin/log?service=cmsDbBrowser&timestamp=1622108680) and reproduced locally with:
```
getPayloadData.py \
--plugin pluginSiPixelLorentzAngle_PayloadInspector \
--plot plot_SiPixelLorentzAngleValuesBarrelCompareTwoTags \
--tag SiPixelLorentzAngleSim_phase1_BoR3_HV350_Tr2000 \
--tagtwo SiPixelLorentzAngle_phase1_EEoR3_HV800_Tr2000  \
--time_type Run \
--iovs '{"start_iov": "1", "end_iov": "1"}' \
--iovstwo '{"start_iov": "1", "end_iov": "1"}' \
--db Prod \
--test ;
```
resulting in:
```
Traceback (most recent call last):
  File "/cvmfs/cms-ib.cern.ch/nweek-02686/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-06-20-0000/bin/slc7_amd64_gcc900/getPayloadData.py", line 8, in <module>
    import yaml
ImportError: No module named yaml
```

Looks related to recent python3 migration changes.
cc:
@francescobrivio @ggovi

**Comment by francescobrivio** [comment]
The error can also be reproduced with
```
cmsrel CMSSW_12_0_0_pre3
cd CMSSW_12_0_0_pre3/src
cmsenv
python
>>> import yaml
```
which gives:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named yaml
```
Same commands in CMSSW_12_0_0_**pre2** work fine.

**Comment by smuzaffar** [comment]
Dueto python3 migration, we have dropped many python2 packages ( all of them are available forpython3 though). Can you please use `python3` to run  `getPayloadData.py` ? Do we have any unit tests to run `getPayloadData.py` ? If yes then it shoudl work with python3 as all tests run for PY3 IBs where `python` is alias to `python3`

**Comment by mmusich** [comment]
Thanks @smuzaffar .
 > Can you please use python3 to run getPayloadData.py ?

Does  this mean basically to change the shebang here:

https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/CondCore/Utilities/scripts/getPayloadData.py#L1
?

> Do we have any unit tests to run getPayloadData.py

No, but I'll add one.

**Comment by smuzaffar** [comment]
yes please change the shebang in the script

**Comment by makortel** [comment]
assign db

**Comment by mmusich** [comment]
with this change-set:
```diff
diff --git a/CondCore/Utilities/scripts/getPayloadData.py b/CondCore/Utilities/scripts/getPayloadData.py
index dd50a321ef3..f01334d6fa3 100755
--- a/CondCore/Utilities/scripts/getPayloadData.py
+++ b/CondCore/Utilities/scripts/getPayloadData.py
@@ -1,5 +1,4 @@
-#!/usr/bin/env python
-
+#!/usr/bin/env python3
 
 from __future__ import print_function
 import shutil
@@ -283,9 +282,9 @@ if __name__ == '__main__':
             try:
                 filename = json.loads( result )['file']
                 #print 'File name',filename
-            except ValueError, e:
+            except ValueError as e:
                 os.write( 2, 'Value error when getting image name: %s\n' % str( e ))
-            except KeyError, e:
+            except KeyError as e:
                 os.write( 2, 'Key error when getting image name: %s\n' % str( e ))
 
             if not filename or not os.path.isfile( filename ):
@@ -294,7 +293,7 @@ if __name__ == '__main__':
             try:
                 with open( filename, 'r' ) as f:
                     shutil.copyfileobj( f, sys.stdout )
-            except IOError, e:
+            except IOError as e:
                 os.write( 2, 'IO error when streaming image: %s' % str( e ))
             finally:
                 os.remove( filename )
``` 

I still get this error message:
```
Traceback (most recent call last):
  File "/tmp/musich/CMSSW_12_0_X_2021-06-20-2300/bin/slc7_amd64_gcc900/getPayloadData.py", line 16, in <module>
    import pluginCondDBV2PyInterface
ImportError: dynamic module does not define module export function (PyInit_pluginCondDBV2PyInterface)
```

which looks related to https://github.com/cms-sw/cmssw/pull/34026

**Comment by davidlange6** [comment]
more likely its that boost python is python2  - so likely we can't move this tool to python3 until boost python does - what else is needed aside from pyyaml?

**Comment by mmusich** [comment]
If it helps that's the list of import statements:

```python
from __future__ import print_function
import shutil
import glob
import json
import yaml
import sys
import sys
import os

from importlib  import import_module
from argparse   import ArgumentParser, RawTextHelpFormatter


import pluginCondDBV2PyInterface
pluginCondDBV2PyInterface.CMSSWInit()
```

**Comment by davidlange6** [comment]
Yep - thanks! I think only yaml has been dropped, but I’ll double check tomorrow.

**Comment by davidlange6** [comment]
Just to follow up - getPayloadData relies on pluginCondDBV2PyInterface which is from

BOOST_PYTHON_MODULE(pluginCondDBV2PyInterface) 

So indeed, this is a boost-python module (so python2 at this moment)

**Comment by mmusich** [comment]
So, what is the plan to restore the functionality? 
Is pyyaml for python2 going to be restored until boost-python is not available for python3?

**Comment by smuzaffar** [comment]
yes, @davidlange6 has already opened https://github.com/cms-sw/cmsdist/pull/7054 to restore it.

**Comment by ggovi** [comment]
Boost-python can ( and should ) be replaced by PyBind11. <use name="py3-pybind11> in the BuildFile should provide the support for python3?  Currently I'm getting the error:

[govi@cmsdev23 src]$ python3
Python 3.8.2 (default, Jun  1 2021, 09:03:26) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pluginCondDBPyBind11Interface
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dynamic module does not define module export function (PyInit_pluginCondDBPyBind11Interface)

Any suggestion on how to fix it?

**Comment by davidlange6** [comment]
it depends on which python you used as a dependency in the so. (pybind11 itself does not depend on python*)

**Comment by ggovi** [comment]
The .cc creating the .so only contains #include <pybind11/pybind11.h>. Here is the BuildFile:

<library file="CondDBPyBind11Wrappers.cc" name="CondDBPyBind11Interface">
  <use name="CondCore/CondDB"/>
  <use name="py3-pybind11"/>
  <use name="python"/>
</library>

**Comment by davidlange6** [comment]
use name="python3"

**Comment by smuzaffar** [comment]
@davidlange6 , this plugin is generated here https://github.com/cms-sw/cmssw/blob/master/CondCore/CondDB/plugins/BuildFile.xml#L6 . It is an edm plugin. I don't think that we can have python3 and python2 based plugins.

**Comment by davidlange6** [comment]
sorry, I was thinking it was this one https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/CondCore/CondDB/plugins/CondDBPyWrappers.cc

(aack)

I would assume that you are correct that we can not mix python2 and python3 plugins (but I haven't tested that). 

Anyway, it looks like we should keep all of this as python2 until a last Big Bang change over - but meanwhile it would be useful that @ggovi confirms that the py3 Its are working ok for him.

**Comment by ggovi** [comment]
@davidlange6 thanks, the fix for the pybind11 works. Now, for the issue reported here I planned in principle create a new plugin based on python3, leaving the one used by conddb in python2. Do I understand correctly that this is not possible?

**Comment by davidlange6** [comment]
I’m not sure - definitely if both get loaded into the same process you’ll have problems. There may be other issues given that the fwk uses python2 (but that is outside of the plugins themselves). I’d suggest just trying it to see what happens

**Comment by smuzaffar** [comment]
@davidlange6 , @ggovi , `scram` uses `edmPluginRefresh` during the build phase to generate edmplugin cache.  `edmPluginRefresh` loads all plugins in once process and having both python2 and 3 plugins will break the build process. I copied different plugisn build for 12.0.X and 12.0.PY3.X in one directory,  ran `edmPluginRefresh` on it and as expacted it failed.

```
> edmPluginRefresh -p lib/slc7_amd64_gcc900
Caught exception An exception of category 'PluginLibraryLoadError' occurred.
Exception Message:
unable to load lib/slc7_amd64_gcc900/pluginAlignPCLThresholds_PayloadInspector.so because libboost_python38.so.1.75.0: cannot open shared object file: No such file or directory
 will ignore pluginAlignPCLThresholds_PayloadInspector.so and continue.
```

**Comment by smuzaffar** [comment]
@ggvi, `CondDBPyBind11Interface` is not really a edm plugin. I would suggest to do the following 
- add the folloiwng at https://github.com/cms-sw/cmssw/blob/master/CondCore/CondDB/plugins/BuildFile.xml#L7 
```
<flags EDM_PLUGIN="0"/>
```
and replace `<use name="python"/>` with `<use name="python3"/>` .This will generate a shared library `libCondDBPyBind11Interface.so` linked with python3.  As this shared lib is generated from plugins directory, so no one should be able to link against it.

- change the name to `CondDBPyBind11Interface` at https://github.com/cms-sw/cmssw/blob/master/CondCore/CondDB/plugins/CondDBPyBind11Wrappers.cc#L32

**Comment by ggovi** [comment]
@smuzaffar Ok, thanks, i'll follow this recipe!

**Comment by mmusich** [comment]
this can be signed and closed

**Comment by smuzaffar** [comment]
closing it as this has been fixed via #34444
